python/mxnet/contrib/nn/layer.py

```diff
@@ -0,0 +1,355 @@
+# coding: utf-8
+# pylint: disable= arguments-differ
+"""Neural network layers."""
+
+from ... import symbol, ndarray
+from ...symbol import Symbol
+from ...ndarray import NDArray
+from .parameter import Parameter, ParameterDict
+
+
+class Layer(object):
+    """Base class for all neural network layers and models.
+
+    Your models should subclass this class.
+
+    Layers can also contain other Layers, allowing you to nest them in a tree
+    structure. You can assign sublayers as regular attributes::
+        from mxnet import nn
+        class Net(nn.Layer):
+            def __init__(self, prefix=None, params=None):
+                super(Net, self).__init__(prefix=prefix, params=params)
+                self.dense1 = nn.Dense(20, in_units=10, prefix=self.prefix+'dense1_')
+                self.dense2 = nn.Dense(20, in_units=20, prefix=self.prefix+'dense2_')
```

Comment:
pytorch use metaclass magic to automatically name the sub-layer according to the variable name it get's assigned. So
```
self.dense1 = nn.Dense(...)
```
will automatically have a (nested) name `dense1`. However, I'm not sure this is actually a feature we want to have. It sometimes could confuse people. If you do something like
```
self.layers = [nn.Dense(...) for i in range(5)]
```
the meta-class will no longer be able to figure it out automatically, and not even be able to collect all the parameters in a layer.